### ISSUE TITLE
Android M (6) compatibility

### ISSUE THREAD

**Issue by EddyVerbruggen** [EddyVerbruggen/Calendar-PhoneGap-Plugin#255]
Add methods to request permission for read and write operations when building with Android SDK 23+.

**Comment by robinbonnes** [comment]
Hello Eddy,

Still getting some compile errors while building for Android:

:compileDebugJava
/platforms/android/src/nl/xservices/plugins/Calendar.java:148: error: cannot find symbol
      if (PackageManager.PERMISSION_GRANTED != ContextCompat.checkSelfPermission(this.cordova.getActivity(), type)) {
                                                            ^
  symbol:   method checkSelfPermission(Activity,String)
  location: class ContextCompat
/platforms/android/src/nl/xservices/plugins/Calendar.java:157: error: cannot find symbol
      ActivityCompat.requestPermissions(
                    ^
  symbol:   method requestPermissions(Activity,String[],int)
  location: class ActivityCompat
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
Note: Some input files use unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.
2 errors
:compileDebugJava FAILED

FAILURE: Build failed with an exception.

Is this related to Android M?

**Comment by EddyVerbruggen** [comment]
Those errors are related to a missing support library on your machine, which the plugin attempts to add: https://github.com/EddyVerbruggen/Calendar-PhoneGap-Plugin/blob/master/plugin.xml#L66

Before moving to other solutions, can you please try to add and run this plugin in a new CLI project first?

**Comment by robinbonnes** [comment]
Thanks for the fast response.

I've tried creating a new CLI project and the same error occurs.
I noticed this is due the plugin is installing version 4.4.4 and not 4.4.5. Any idea why it's fetching the old version?

I tried both:
cordova plugin add cordova-plugin-calendar
cordova plugin add https://github.com/EddyVerbruggen/Calendar-PhoneGap-Plugin

Should I add the --use-git parameter or #master ?

**Comment by EddyVerbruggen** [comment]
4.4.5 wasn't on npm yet - can you please retry? Also make sure you're using a recent Cordova CLI version.

**Comment by henkkelder** [comment]
Same problem on PGB. Version there is 4.4.4 and my app won't build.

**Comment by EddyVerbruggen** [comment]
Can you try using gradle and a recent phonegap version on pgb? I should probably add that to the readme.

**Comment by henkkelder** [comment]
Maakt niet uit Eddy.

 

Cli 5.2.0 bij PGB geeft de fout

Error - Some official plugins have to be updated if using PhoneGap >= 4.0.0. Please upgrade the version of any plugins that may include the following file: Calendar.java 

 

Maar ook een lokale build geeft een fout Cordova 5.4.1, plugin versie 4.4.5. Zie hieronder.

 

En als ik niet de ‘support repository’ installeer krijg ik weer andere foutmeldingen.

 

Gr

Henk Kelder

=====================================================================================

 

D:\opwekking2014\platforms\android\src\nl\xservices\plugins\Calendar.java:148: 

error: cannot find symbol

      if (PackageManager.PERMISSION_GRANTED != ContextCompat.checkSelfPermission

(this.cordova.getActivity(), type)) {

                                                            ^

  symbol:   method checkSelfPermission(Activity,String)

  location: class ContextCompat

D:\opwekking2014\platforms\android\src\nl\xservices\plugins\Calendar.java:157: error: cannot find symbol

      ActivityCompat.requestPermissions(

                    ^

  symbol:   method requestPermissions(Activity,String[],int)

  location: class ActivityCompat

Note: Some input files use or override a deprecated API.

Note: Recompile with -Xlint:deprecation for details.

2 errors

:compileDebugJava FAILED

**Comment by EddyVerbruggen** [comment]
@henkkelder (English please, thanks) - did you configure your project to build with Gradle? This project build fine on PGB for Android:

Calendar plugin:
https://github.com/EddyVerbruggen/X-Services-PhoneGap-Build-Plugins-Demo/blob/master/config.xml#L34

Gradle config:
https://github.com/EddyVerbruggen/X-Services-PhoneGap-Build-Plugins-Demo/blob/master/config.xml#L50
```

**Comment by henkkelder** [comment]
Yes I did. The following line is in my config.xml:

 

  <preference name="android-build-tool" value="gradle" />

 

But this shouldn’t be necessary:

 

http://phonegap.com/blog/2015/09/28/android-using-gradle/

 

 

But local builds (using CLI on Windows) also fail whenever the plugin is included with the errormessage as in my previous message.

 

Groeten

Henk

**Comment by pyroon** [comment]
Hi
Any update on this topic? I'm facing the same issue (with android-build-tool set)

Cheers,
Chris

**Comment by EddyVerbruggen** [comment]
If you're building locally then you should run the SDK manager and update to the latest bits.

**Comment by pyroon** [comment]
I'm running on the latest version and calendar plugin is on 4.4.5. But I'm getting the compilation errors below:

```
:compileDebugJava/Users/christoph/Documents/workspace/airzone-app/platforms/android/src/nl/xservices/plugins/Calendar.java:148: error: cannot find symbol
      if (PackageManager.PERMISSION_GRANTED != ContextCompat.checkSelfPermission(this.cordova.getActivity(), type)) {
                                                            ^
  symbol:   method checkSelfPermission(Activity,String)
  location: class ContextCompat
/Users/christoph/Documents/workspace/airzone-app/platforms/android/src/nl/xservices/plugins/Calendar.java:157: error: cannot find symbol
      ActivityCompat.requestPermissions(
                    ^
  symbol:   method requestPermissions(Activity,String[],int)
  location: class ActivityCompat
```

**Comment by EddyVerbruggen** [comment]
Please update your android support lib to the latest version, clean everything, etc, since to me that `ActivityCompat` is clearly an old version.

**Comment by apstory** [comment]
Hi, please can you assist, I'm still getting the below error even after updating android support lib to version 23.1.1. I've also cleaned  everything I could think of:

:compileDebugJava
C:\Source\Repos\...\platforms\android\src\nl\xservices\plugins\Calendar.java:148: error: cannot find symbol
      if (PackageManager.PERMISSION_GRANTED != ContextCompat.checkSelfPermission(this.cordova.getActivity(), type)) {
                                                            ^
  symbol:   method checkSelfPermission(Activity,String)
  location: class ContextCompat
C:\Source\Repos\...\platforms\android\src\nl\xservices\plugins\Calendar.java:157: error: cannot find symbol
      ActivityCompat.requestPermissions(
                    ^
  symbol:   method requestPermissions(Activity,String[],int)
  location: class ActivityCompat
Note: Some input files use or override a deprecated API.
Note: Recompile with -Xlint:deprecation for details.
2 errors
:compileDebugJava FAILED

**Comment by newagesolution** [comment]
I am using "cordova-cli": "5.4.1". having same issue. Would you be able to help?

**Comment by EddyVerbruggen** [comment]
Can you try setting the target sdk to 23 in AndroidManifest.xml and make sure the support libraries are up to date? If others have better ideas I'd love to hear them.

**Comment by newagesolution** [comment]
First I had to install all things related to sdk 23 using SDK Manager. Basically I updated all the sdk using SDK Manager. I was able to compile. Thank you.

**Comment by jomarocas** [comment]
for me working in a moto x 2 gen, but my issue is when i touch in open calendar in app test, always close and i need restart the smartphone for working again, but when i make open calendar, again close the app and no working more, i test the app in samsung galaxy s4 and works good, thanks if any have the solution

**Comment by jinjin86** [comment]
I added below code to gradle and then it works for me.

dependencies {
...
compile 'com.android.support:support-v4:23.1.0'
..
}

**Comment by misterschwamm** [comment]
Hi, i still have this error. I have updated everything related to sdk 23 and the support library and have build with android-23. Does someone have a solution?
Thanks!

**Comment by animista01** [comment]
I've the same error that @jomarocas has, my Android is a Moto G Turbo with Android M, The first time I could create a event successfully in my Calendar app, but then the calendar app crash, it can't start, can't even be open. I deleted my app, even deleted the calendar app and downloading it again but the app won't start.

**Comment by jclark-dot-org** [comment]
Possible solution for those still having this issue:  check platform/android/libs for "android-support-v13.jar".  If found, remove it, and find any plugin(s) referencing it in their plugin.xml, and remove the reference.  Details below.

I was having this issue, despite checking my installed libraries multiple times.  With the help of debugging output (cordova build android -d), I was able to figure out that one of my plugins (com.phonegap.plugins.PushPlugin) had a dependency on "android-support-v13.jar" in its plugin.xml:

<source-file src="src/android/com/plugin/android-support-v13.jar" target-dir="libs/" />

This resulted in a copy of android-support-v13.jar in my platforms/android/libs directory.  Apparent "v13" was a special version of the v4 Support library that shipped as part of rev 3 back in July 2011 and offered API 13 compatibility.  I fixed this by commenting out the source-file in pushplugin's plugin.xml, and deleting the jar from platforms/android/libs.  Now the build is working again.
